pkg/cache/v3/linear.go

```diff
@@ -29,37 +30,86 @@ import (
 // cachedResource is used to track resources added by the user in the cache.
 // It contains the resource itself and its associated version (currently in two different modes).
 type cachedResource struct {
-	types.Resource
+	name string
+
+	resource types.Resource
+	ttl      *time.Duration
 
 	// cacheVersion is the version of the cache at the time of last update, used in sotw.
 	cacheVersion string
+
 	// stableVersion is the version of the resource itself (a hash of its content after deterministic marshaling).
 	// It is lazy initialized and should be accessed through getStableVersion.
 	stableVersion string
+
+	// marshaledResource contains the marshaled version of the resource.
+	// It is lazy initialized and should be accessed through getMarshaledResource
+	marshaledResource []byte
+
+	// mu is the mutex used to lazy compute the marshaled resource and stable version.
+	mu sync.Mutex
+}
+
+func newCachedResource(name string, res types.Resource, cacheVersion string) *cachedResource {
+	return &cachedResource{
+		name:         name,
+		resource:     res,
+		cacheVersion: cacheVersion,
+	}
 }
 
-func newCachedResource(res types.Resource, cacheVersion string) *cachedResource {
+func newCachedResourceWithTTL(name string, res types.ResourceWithTTL, cacheVersion string) *cachedResource {
 	return &cachedResource{
-		Resource:     res,
+		name:         name,
+		resource:     res.Resource,
+		ttl:          res.TTL,
 		cacheVersion: cacheVersion,
 	}
 }
 
+// getMarshaledResource lazily marshals the resource and returns the bytes.
```

Comment:
When you say lazily you mean returns the cached resource right if it exits, otherwise it marshals it?

When I think of lazily returns i think of haskell style lazy returns.

Reply:
I'm not familiar with Haskell, but here it is lazy computed within a getter. 
In this case it's painfully returning an error, which can trigger if the user passes in a type not known from the protoregistry

Reply:
https://wiki.haskell.org/index.php?title=Lazy_evaluation